### PR TITLE
MCO-417: MCO-418: MCD watches/create/update password with MachineConfig

### PR DESCRIPTION
**Pull request by dkhater-redhat**
**- What I did**
Created a SetPasswordHash function to allow a user to create a password; MCD learns to make/update a password for the user core based on what is provided by MachineConfig.

**- How to verify it**
Set a password for the user core by adding a machine config that specifies the password (following ignition format). Shell into the pod and log in as core (not from root, or the password option will not prompt), verifying that your password works.

**- Description for the changelog**
SetPassword Hash allows users to create and update their password for core (password cannot be empty); edited verifyUserFields allows the user field verification if the password or ssh key exists.

### CODE REVIEW

File: pkg/daemon/update.go
Comment (jkyros) [inline pkg/daemon/update.go:590]: I think this defer got accidentally removed -- we do need it, and we also will need a matching one for the password so the password gets reverted if we hit an error.

**Review by jkyros — COMMENTED** [review]
Just a couple of things I noticed while scrolling through, I'll do a second pass later.

File: pkg/daemon/update.go
Comment (dkhater-redhat) [reply pkg/daemon/update.go:590]: yes i accidentally removed this- thanks john!

**Comment by rioliu-rh** [comment]
this is 4.13 new feature, please hold this PR for QE verification when dev's review is done.  thanks

**Comment by sinnykumari** [comment]
Thanks for noticing this Rio! putting hold so that after getting reviewed by dev team, qe team can verify the feature and add qe_approved label as part of pre-merge criteria
/hold

File: pkg/daemon/update.go
Comment (sinnykumari) [inline pkg/daemon/update.go:586]: I think you meant here "error rolling back SSH keys updates"?

File: pkg/daemon/update.go
Comment (sinnykumari) [inline pkg/daemon/update.go:599]: Should this defer call SetPasswordHash() ?
Looks to me that ordering is a bit messed up for update SSH Key and set  password

**Review by sinnykumari — COMMENTED** [review]
It will be good to have unit test case to make sure that we are only creating and updating core user password. Like we have today for ssh keys https://github.com/openshift/machine-config-operator/blob/master/pkg/daemon/update_test.go#L299

File: pkg/daemon/update.go
Comment (sinnykumari) [inline pkg/daemon/update.go:941]:
```suggestion
			return fmt.Errorf("ignition passwd user section contains unsupported changes. Supported changes includes Password Hash and SSH key")
```

File: pkg/daemon/update.go
Comment (sinnykumari) [inline pkg/daemon/update.go:1838]: don't understand the need of this comment. Are we planning to mock this?

File: pkg/daemon/update.go
Comment (sinnykumari) [inline pkg/daemon/update.go:1851]: nit: newline not needed here.

**Review by sinnykumari — COMMENTED** [review]
Added few comments. To avoid confusion, we will need to update existing comments where we just say that only sshkey update is supported for user -> supporting updating ssh key  and password hash
Overall looks good. Nice work!

File: pkg/daemon/update.go
Comment (jkyros) [inline pkg/daemon/update.go:1853]: I think this is probably why we get the panic -- deref before nil check. 

I also think the line numbers probably don't match up to the panic because of how the test merges the code when it builds the PR: https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift_machine-config-operator/3539/pull-ci-openshift-machine-config-operator-master-e2e-gcp-op/1623093773879742464/clone-log.txt

File: pkg/daemon/update.go
Comment (dkhater-redhat) [reply pkg/daemon/update.go:1838]: this was taken from the update ssh keys function- I believe that chunk of code was to mock the nodes existence to ensure it runs if a node exists

**Comment by sinnykumari** [comment]
gofmt related changes should be already fixed in master branch. To avoid confusion, I will recommend two things:
1. Rebase branch with latest master branch
2. Squash commits into single commit or if needed multiple commits that are logically separated so that git commit history is well reflected

File: pkg/daemon/update.go
Comment (sinnykumari) [inline pkg/daemon/update.go:1816]: Curious, any reason to do both ssh key and password update in same function?  Keeping them separate would be less error prone because handling defer call during failure will not be needed for both.

Comment (dkhater-redhat) [reply pkg/daemon/update.go:1816]: Yes- it was because the cyclomatic complexity was over limit in (*Daemon).update. 
So I decided to combine it with updateSSHKeys because the functions were similar (they both mock nodes and they both access the passwd user struct).

Comment (dkhater-redhat) [reply pkg/daemon/update.go:1816]: I originally kept it separate but the cyclomatic complexity was already at limit when I started working on this.

Comment (jkyros) [reply pkg/daemon/update.go:1816]: If we keep the mushed together, we at least need to make sure the passwordhash section is behind the mock (because usermod can't run if there's no node to run against) and the name should reflect what it does (e.g. `UpdatePasswdUsers` or `UpdatePasswdAndSshKeys` or something like that)

Yes, the coreos user check would be duplicated if we split them back out, but that could be a `coreUserExists()` function if we were worried? 

Honestly I'd be okay with just setting `//nolint:gocyclo` on `update()` and leaving UpdatePasswd and UpdateSSH as  separate since the whole point of this is clarity? 

And I feel like that would be clearer than a middle ground like: 

```
func UpdatePasswdUsers() { 

	updatePasswd()
	defer updatePasswd()
	
	updateSSHKeys
	defer updateSSHKeys() 
	
}
```
but that's maybe a matter of opinion?

Comment (dkhater-redhat) [reply pkg/daemon/update.go:1816]: I'm ok with both opinions- not sure what the repercussions of setting `//nolint:gocyclo` on `update()` will do. If its ok then I will be happy to revert because it is more explicit as two separate functions

Comment (sinnykumari) [reply pkg/daemon/update.go:1816]: Agree with John :)

**Comment by rioliu-rh** [comment]
Do we need to test it on RHEL node?

**Comment by dkhater-redhat** [comment]
> Do we need to test it on RHEL node?

to my understanding, we can test on a RHEL node because it probably works and would be supported. The only foreseeable issue is that RHEL supports different passwords than RHCOS; however, we could be specific and document what it supports.

**Comment by rioliu-rh** [comment]
> > Do we need to test it on RHEL node?
> 
> to my understanding, we can test on a RHEL node because it probably works and would be supported. The only foreseeable issue is that RHEL supports different passwords than RHCOS; however, we could be specific and document what it supports.

thanks @dkhater-redhat , a little difference on RHEL,  user `core` is not created by default on RHEL node. we need to create it manually @sergiordlr can provide more details

**Comment by sinnykumari** [comment]
It is indeed a good idea to test on RHEL node for at least these two cases:
1. User core exist on a RHEL node  and MachineConfig with passwordHash has been applied in the cluster
2. User core doesn't exist on a RHEL node and MachineConfig with passwordHash has been applied in the cluster

File: pkg/daemon/update.go
Comment (sinnykumari) [inline pkg/daemon/update.go:1602]: nit: this comment can be removed as we are not doing any mock while performing passwd update

File: pkg/daemon/update.go
Comment (sinnykumari) [inline pkg/daemon/update.go:909]: It looks to me that verfiyUserField may fail (as else condition will run) in a situation  when rendered config have user core && PasswordHash != nil && SSHAuthorizedKeys  == nil

File: pkg/daemon/update.go
Comment (sinnykumari) [inline pkg/daemon/update.go:916]: This can be changed to "SSH keys and password hash are  reconcilable"

File: pkg/daemon/update.go
Comment (sinnykumari) [inline pkg/daemon/update.go:917]: error messaging in else condition needs to be updated based on how we fix comment  https://github.com/openshift/machine-config-operator/pull/3539/files#r1106978031 .
Note that with passwordhash support, user  can specify any of these:
1. ssh key for user core
2. password hash for user core
3. ssh key and password hash for user core

File: pkg/daemon/update.go
Comment (dkhater-redhat) [reply pkg/daemon/update.go:909]: This section was written by someone else- how should i resolve this?

Comment (sinnykumari) [reply pkg/daemon/update.go:909]: As discussed offline, you are free to make changes anywhere in MCO repo :) Also, it is good to fix any known regression(as a result of the change introduced) before merging PR

File: pkg/daemon/update.go
Comment (sinnykumari) [inline pkg/daemon/update.go:905]:
```suggestion
// outside of SSHAuthorizedKeys and passwordHash.
```

**Review by sinnykumari — APPROVED** [review]
One minor nit, otherwise looks good.

**Comment by sinnykumari** [comment]
Nice work Dalia, this looks great. Will let @jkyros  or any other team members take a final look and add lgtm.

**Review by cheesesashimi — CHANGES_REQUESTED** [review]
Nice job @dkhater-redhat! I have two minor requests, although neither pertains to the code, but rather to the PR:
1. Edit the tag line (`MCO-417: MCO-418:`) to be more descriptive. You can use the same text you used for your commit. Your edit should look something like `MCO-417: Your Description Here` so the OpenShift CI bot maintains the association between this PR and the JIRA card.
2. Fill in the "What I did", "How to verify it", and "Description for the changelog" fields with a brief description of what this PR does. This PR is pretty straightforward, so you don't have to put much in those fields.

The reason I'm requesting these things is because:

- The PR and the code it contains codifies what was done and how it was done. Adding those things helps codify the "why" behind it so that you and others can refer to it in the future.
- The OpenShift release automation consumes this information and when looking at the release notes for a particular release, it's helpful to see something like, "Enables support for core user password hashes".

**Comment by jkyros** [comment]
I think this looks good and am going to lgtm it, buuut I think we also need to hold this for pre-merge QE, don't we @rioliu-rh ? 
/hold

File: pkg/daemon/update.go
Comment (dkhater-redhat) [reply pkg/daemon/update.go:914]: yes I agree with you! will commit

Comment (jkyros) [reply pkg/daemon/update.go]: lol looking back I think Sinny already had you change that. Ignore me I guess :smile:

**Comment by jkyros** [comment]
it was already held. Prow is on break again I guess. :smile:
/lgtm

**Comment by cheesesashimi** [comment]
I worked with @dkhater-redhat to add the testing commits to this PR. While trying to debug the e2e test, we added additional unit test cases to target both the `MergeMachineConfigs()` and `newMachineConfigDiff()` functions to validate that the existing code paths would detect changes to the `PasswordHash` field as well as merge them correctly. These unit tests proved that the `MergeMachineConfigs()` and `newMachineConfigDiff()` functions handled those cases satisfactorily and without modification.

**Comment by cheesesashimi** [comment]
Also, just for posterity, we discovered that the `TestNoReboot` e2e test is dependent upon at least one other test within the MCO e2e test suite. We discovered this because the test assumes the `infra` MachineConfigPool is already present when it shouldn't be. Furthermore, it does not remove the `infra` MachineConfigPool when it is finished. I have some code in https://github.com/openshift/machine-config-operator/pull/3558 which could be used to modify the `TestNoReboot` test case in the future to remove that dependency. This is definitely a future TODO and not something we plan to accomplish within this PR.

**Comment by cheesesashimi** [comment]
After seeing the successful e2e test run in e2e-gcp-op as well as a successful unit test run:

/lgtm
/approved

This is now ready for QE verification.

**Review by cheesesashimi — APPROVED** [review]

**Comment by sinnykumari** [comment]
Removing hold as this has been verified by QE
/hold cancel

**Pull request merged**
